libdrgn/kallsyms.c

```diff
@@ -0,0 +1,659 @@
+
+	// Now find the end of the names array by skipping through it, then copy
+	// that into host memory.
+	names_idx = 0;
+	kr->long_names = guess_long_names(prog);
+	for (size_t i = 0; i < kr->num_syms; i++) {
+		err = drgn_program_read_u8(prog,
+					   loc->kallsyms_names + names_idx,
+					   false, &len_u8);
+		if (err)
+			return err;
+		len = len_u8;
+		if ((len & 0x80) && kr->long_names) {
+			err = drgn_program_read_u8(prog,
+						loc->kallsyms_names + names_idx + 1,
+						false, &len_u8);
+			if (err)
+				return err;
+			// 73bbb94466fd3 ("kallsyms: support "big" kernel
+			// symbols") mentions that ULEB128 is used, but only
+			// implements the ability to encode lengths with 2
+			// bytes, for a maximum value of 16k. It's possible in
+			// the future we may need to support larger sizes, but
+			// it's difficult to predict the future of the kallsyms
+			// format. For now, just check that there's no third
+			// byte to the length.
+			if (len_u8 & 0x80)
+				return drgn_error_format(
+					DRGN_ERROR_OTHER,
+					"Unexpected 3-byte length encoding in kallsyms names"
+				);
+			len = (len & 0x7F) | (len_u8 << 7);
+			names_idx++;
+		}
+		if (__builtin_add_overflow(names_idx, len + 1, &names_idx))
+			return drgn_error_format(
+				DRGN_ERROR_OTHER, "kallsyms: overflowed names_idx");
+	}
+	kr->names_len = names_idx;
+	kr->names = malloc(names_idx);
+	if (!kr->names)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->names,
+				       loc->kallsyms_names,
+				       names_idx, false);
+	if (err)
+		return err;
+
+	return NULL;
+}
+
+/**
+ * Extract the symbol name and type
+ * @param kr Registry containing kallsyms data
+ * @param offset Starting index within names array for this symbol
+ * @param sb Buffer to write output symbol to
+ * @param[out] kind_ret Where to write the symbol kind data
+ * @returns The offset of the next symbol in the names array
+ */
+static struct drgn_error *
+kallsyms_expand_symbol(struct kallsyms_reader *kr, uint32_t *offset,
+		       struct string_builder *sb, char *kind_ret)
+{
+	if (*offset >= kr->names_len)
+		return drgn_error_format(DRGN_ERROR_OTHER,
+					 "kallsyms: invalid offset %" PRIu32 ", max offset is %zu",
+					 *offset, kr->names_len);
+	uint8_t *data = &kr->names[*offset];
+	unsigned int len = *data;
+	bool skipped_first = false;
+
+	// Bounds checking: if we've overrun the buffer, we'll return an error
+	// below when we bounds check *offset + len + 1
+	if ((len & 0x80) && kr->long_names && *offset + 1 < kr->names_len) {
+		data++;
+		(*offset)++;
+		len = (0x7F & len) | (*data << 7);
+	}
+
+	// Bounds check the bytes we will read.
+	*offset += len;
+	if (*offset >= kr->names_len)
+		return drgn_error_format(DRGN_ERROR_OTHER,
+					 "kallsyms: invalid offset %" PRIu32 ", max offset is %zu",
+					 *offset, kr->names_len);
+	// Leave offset pointing at the next token, which we will bounds check
+	// on the next call.
+	*offset += 1;
```

Comment:
I should've mentioned this earlier, but `struct binary_buffer` can save you from doing a lot of this tricky bounds checking manually. Something like this passes tests, but please look it over and update the `kallsyms_expand_symbol` comment:

```diff
diff --git a/libdrgn/kallsyms.c b/libdrgn/kallsyms.c
index 5057dffc..1e939959 100644
--- a/libdrgn/kallsyms.c
+++ b/libdrgn/kallsyms.c
@@ -4,6 +4,7 @@
 #include <ctype.h>
 #include <stddef.h>
 
+#include "binary_buffer.h"
 #include "kallsyms.h"
 #include "program.h"
 #include "drgn_internal.h"
@@ -221,6 +222,14 @@ kallsyms_copy_tables(struct drgn_program *prog, struct kallsyms_reader *kr,
 	return NULL;
 }
 
+static struct drgn_error *kallsyms_binary_buffer_error(struct binary_buffer *bb,
+						       const char *pos,
+						       const char *message)
+{
+	return drgn_error_format(DRGN_ERROR_OTHER,
+				 "couldn't parse kallsyms: %s", message);
+}
+
 /**
  * Extract the symbol name and type
  * @param kr Registry containing kallsyms data
@@ -230,36 +239,22 @@ kallsyms_copy_tables(struct drgn_program *prog, struct kallsyms_reader *kr,
  * @returns The offset of the next symbol in the names array
  */
 static struct drgn_error *
-kallsyms_expand_symbol(struct kallsyms_reader *kr, uint32_t *offset,
+kallsyms_expand_symbol(struct kallsyms_reader *kr,
+		       struct binary_buffer *names_bb,
 		       struct string_builder *sb, char *kind_ret)
 {
-	if (*offset >= kr->names_len)
-		return drgn_error_format(DRGN_ERROR_OTHER,
-					 "kallsyms: invalid offset %" PRIu32 ", max offset is %zu",
-					 *offset, kr->names_len);
-	uint8_t *data = &kr->names[*offset];
-	unsigned int len = *data;
+	struct drgn_error *err;
+
+	uint64_t len;
+	err = binary_buffer_next_uleb128(names_bb, &len);
+	if (err)
+		return err;
+	const uint8_t *data = (uint8_t *)names_bb->pos;
+	err = binary_buffer_skip(names_bb, len);
+	if (err)
+		return err;
+
 	bool skipped_first = false;
-
-	// Bounds checking: if we've overrun the buffer, we'll return an error
-	// below when we bounds check *offset + len + 1
-	if ((len & 0x80) && kr->long_names && *offset + 1 < kr->names_len) {
-		data++;
-		(*offset)++;
-		len = (0x7F & len) | (*data << 7);
-	}
-
-	// Bounds check the bytes we will read.
-	*offset += len;
-	if (*offset >= kr->names_len)
-		return drgn_error_format(DRGN_ERROR_OTHER,
-					 "kallsyms: invalid offset %" PRIu32 ", max offset is %zu",
-					 *offset, kr->names_len);
-	// Leave offset pointing at the next token, which we will bounds check
-	// on the next call.
-	*offset += 1;
-
-	data += 1;
 	while (len) {
 		char *token_ptr = &kr->token_table[kr->token_index[*data]];
 		while (*token_ptr) {
@@ -291,12 +286,14 @@ search_for_string(struct kallsyms_reader *kr, const char *name, ssize_t *ret)
 {
 	STRING_BUILDER(sb);
 	size_t len = strlen(name);
-	uint32_t names_idx = 0;
+	struct binary_buffer names_bb;
+	binary_buffer_init(&names_bb, kr->names, kr->names_len, false,
+			   kallsyms_binary_buffer_error);
 	for (ssize_t i = 0; i < kr->num_syms; i++) {
 		char kind;
 		sb.len = 0;
 		struct drgn_error *err =
-			kallsyms_expand_symbol(kr, &names_idx, &sb, &kind);
+			kallsyms_expand_symbol(kr, &names_bb, &sb, &kind);
 		if (err)
 			return err;
 		if (sb.len == len && strcmp(name, sb.str) == 0) {
@@ -534,13 +531,15 @@ drgn_load_builtin_kallsyms(struct drgn_program *prog,
 	drgn_symbol_index_builder_init(&builder);
 	STRING_BUILDER(sb);
 
-	uint32_t names_idx = 0;
+	struct binary_buffer names_bb;
+	binary_buffer_init(&names_bb, kr.names, kr.names_len, false,
+			   kallsyms_binary_buffer_error);
 	for (int i = 0; i < kr.num_syms; i++) {
 		struct drgn_symbol symbol;
 		char kind;
 		uint64_t size = 0;
 		sb.len = 0;
-		err = kallsyms_expand_symbol(&kr, &names_idx, &sb, &kind);
+		err = kallsyms_expand_symbol(&kr, &names_bb, &sb, &kind);
 		if (err)
 			return err;
 		if (sb.len == 0)
```

Reply:
Wow, that looks like an incredibly convenient API!

Reply:
I reimplemented it myself from the binary buffer docstrings, and came to basically the same code, so I feel good about the correctness. Thanks for this, I'll probably end up redoing some of my BTF code with the binary_buffer for better safety.